### PR TITLE
[27] Set up for relevance classifier

### PR DESCRIPTION
**Pull request by j-gillam**
---

# Description

This PR addresses part of [Issue 27.
](https://github.com/nestauk/discovery_child_development/issues/27)

Following on from PR 31, this PR uses the generalised OpenAlex code to create a pipeline for:

- Pulling in and saving "not relevant" concepts from OpenAlex using broad seed concepts. 
- Preprocessing the openalex data (01_preprocess_openalex_broad.py)
- Embedding the text data (02_embed_openalex_docs_broad.py)
- Creating the training/validation sets (03_binary_classifier_training_data.py)
- Creates getters for the processed "not_relevant" dataset and the training/validation/test sets.

To review this PR:

- Check the getters pull in the correct datasets. 
- Check out the README. Does it make sense? Test the first step with `-- production False` instead of `True`. 
- Check the general code structure. 

(Realised I have spelt classifier wrong in the pipeline! Will fix in a later PR)

# Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [x] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [x] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [x] I have requested a code review

### CODE REVIEW

File: discovery_child_development/pipeline/openalex/binary_classifer/README.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
+# Openalex: Binary Classifier
+
+In this folder you will find the code for the binary classifier used to identify relevant and irrelevant Openalex works.
+
+### Instructions to replicate the pipeline
+
+1. To pull in the openalex metadata for the classifier, run the following scripts in the repository root:
+
+`python discovery_child_development/pipeline/openalex/00_openalex_metaflow.py run --production True --random_sample True --chunk_size 1 --concept_ids openalex_broad_concepts --chunk_number 1`
```
Comment (RFOxbury) [inline discovery_child_development/pipeline/openalex/binary_classifer/README.md]: Maybe I should have checked this on the last PR but I just ran `python discovery_child_development/pipeline/openalex/00_openalex_metaflow.py run --production False --random_sample True --chunk_size 1 --concept_ids openalex_broad_concepts --chunk_number 1` and it gave me the error `Error: no such option: --chunk_number` - is that expected? If so, maybe just change this line. Or maybe something works differently if `production` is `False`

File: discovery_child_development/pipeline/openalex/binary_classifer/03_binary_classifier_training_data.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    # Loading in the data
+    logging.info("Loading in the data...")
+    openalex_data = openalex.get_abstracts()
+    openalex_broad_data = openalex_broad_concepts.get_abstracts_broad()
+
+    # Create test set for the classifier
+    logging.info("Creating test set...")
+    # 50/50 split of EY seed list and broader concepts
+    test_data = pd.concat(
+        [openalex_broad_data.sample(500), openalex_data.sample(500)],
```
Comment (RFOxbury) [inline discovery_child_development/pipeline/openalex/binary_classifer/03_binary_classifier_training_data.py]: Maybe set a seed for this part for reproducibility? I think you should be able to use the [random_state](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sample.html) parameter for `sample()`

File: discovery_child_development/getters/binary_classifier/binary_classifier_training_set.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+CONCEPT_IDS = "|".join(config["openalex_concepts"])
+BROAD_CONCEPT_IDS = "|".join(config["openalex_broad_concepts"])
+YEARS = [str(y) for y in config["openalex_years"]]
+YEARS = "-".join(YEARS)
+
+PATH_FROM = "data/openAlex/processed/binary_classifier/"
+
+
+def get_text_training(
```
Comment (RFOxbury) [inline discovery_child_development/getters/binary_classifier/binary_classifier_training_set.py]: There may be some deduplication needed because some OpenAlex IDs come up multiple times?

I ran this code:
```
from discovery_child_development.getters.binary_classifier import binary_classifier_training_set, binary_classifier_test_set, binary_classifier_validation_set

training_all = binary_classifier_training_set.get_text_training()
training_50 = binary_classifier_training_set.get_text_training(identifier="50")
training_20 = binary_classifier_training_set.get_text_training(identifier="20")

for val in [max(training_all['openalex_id'].value_counts()), max(training_50['openalex_id'].value_counts()), max(training_20['openalex_id'].value_counts())]:
    print(val)
```

File: discovery_child_development/getters/binary_classifier/binary_classifier_test_set.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+def get_text_test(
+    concepts=CONCEPT_IDS, years=YEARS, bucket=S3_BUCKET, path_from=PATH_FROM
+):
+    """Downloads OpenAlex text data (titles and abstracts) from S3. Contains all the data used
+    for the test set for training the binary classifier.
+
+    Args:
+        concepts (str, optional): The concept IDs used in the metaflow. Defaults to CONCEPT_IDS.
+        years (str, optional): The years for which we have data eg "2019_2020". Defaults to YEARS.
+        identifier (str, optional): The identifier for the training data. Defaults to "all". Options are "50", "20", "all".
```
Comment (RFOxbury) [inline discovery_child_development/getters/binary_classifier/binary_classifier_test_set.py]: Could you update this docstring? As `identifier` is not an argument for the test set, right?

File: discovery_child_development/pipeline/openalex/binary_classifer/03_binary_classifier_training_data.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        keys=["not relevant", "relevant"],
```
Comment (RFOxbury) [inline discovery_child_development/pipeline/openalex/binary_classifer/03_binary_classifier_training_data.py:48]: FYI I think using `keys` sets the label to be the index, and as a result the getter for the test set brings back a dataframe with an extra column `Unnamed: 1`

**Review by RFOxbury — APPROVED** [review]
Hello, thanks for this PR! The pipeline scripts look good and I think it's a really good idea to try out sets with different proportions of relevant/non-relevant data.

The only thing I wasn't sure about is that the getters for the training and validation contain some duplicate OpenAlex IDs. It should be simple to get rid of these, or leave them as they are if this was intentional

File: discovery_child_development/pipeline/openalex/binary_classifer/README.md (hunk 1 above)
Comment (j-gillam) [reply discovery_child_development/pipeline/openalex/binary_classifer/README.md]: Ah oops! It's a typo.. chunk_number should not exist, should just be:
`python discovery_child_development/pipeline/openalex/00_openalex_metaflow.py run --production True --random_sample True --chunk_size 1 --concept_ids openalex_broad_concepts`

File: discovery_child_development/pipeline/openalex/binary_classifer/03_binary_classifier_training_data.py (hunk 2 above)
Comment (j-gillam) [reply discovery_child_development/pipeline/openalex/binary_classifer/03_binary_classifier_training_data.py]: Oh yeah definitely! Good spot!

File: discovery_child_development/pipeline/openalex/binary_classifer/README.md (hunk 1 above)
Comment (RFOxbury) [reply discovery_child_development/pipeline/openalex/binary_classifer/README.md]: Thanks, that worked! :)

File: discovery_child_development/getters/binary_classifier/binary_classifier_test_set.py (hunk 4 above)
Comment (j-gillam) [reply discovery_child_development/getters/binary_classifier/binary_classifier_test_set.py]: Yep!

File: discovery_child_development/pipeline/openalex/binary_classifer/03_binary_classifier_training_data.py (hunk 5 above)
Comment (j-gillam) [reply discovery_child_development/pipeline/openalex/binary_classifer/03_binary_classifier_training_data.py:48]: Yeah I realised that on friday! I fixed it but forgot to update the PR woops!

File: discovery_child_development/getters/binary_classifier/binary_classifier_training_set.py (hunk 3 above)
Comment (j-gillam) [reply discovery_child_development/getters/binary_classifier/binary_classifier_training_set.py]: Darn! Yeah I have figured out why that is (because the concept ids are bringing back some of the same results). I will add a deduplication step. Good catch!

**Review by RFOxbury — APPROVED** [review]
Thanks Jess, looks good!
